crates/ty_ide/src/docstring.rs

```diff
@@ -0,0 +1,628 @@
+//! Docstring parsing utilities for language server features.
+//!
+//! This module provides functionality for extracting structured information from
+//! Python docstrings, including parameter documentation for signature help.
+//! Supports Google-style, NumPy-style, and reST/Sphinx-style docstrings.
+//! There are no formal specifications for any of these formats, so the parsing
+//! logic needs to be tolerant of variations.
+
+use regex::Regex;
+use std::collections::HashMap;
+
+/// Extract parameter documentation from popular docstring formats.
+/// Returns a map of parameter names to their documentation.
+pub fn get_parameter_documentation(docstring: &str) -> HashMap<String, String> {
+    let mut param_docs = HashMap::new();
+
+    // Google-style docstrings
+    if let Some(google_docs) = extract_google_style_params(docstring) {
+        param_docs.extend(google_docs);
+    }
+
+    // NumPy-style docstrings
+    if let Some(numpy_docs) = extract_numpy_style_params(docstring) {
+        param_docs.extend(numpy_docs);
+    }
+
+    // reST/Sphinx-style docstrings
+    if let Some(rest_docs) = extract_rest_style_params(docstring) {
+        param_docs.extend(rest_docs);
+    }
+
+    param_docs
+}
+
+/// Extract parameter documentation from Google-style docstrings.
+fn extract_google_style_params(docstring: &str) -> Option<HashMap<String, String>> {
+    let mut param_docs = HashMap::new();
+
+    // Find Args/Arguments/Parameters sections
+    let section_regex = Regex::new(r"(?i)^\s*(Args|Arguments|Parameters)\s*:\s*$").ok()?;
+    let param_regex = Regex::new(r"^\s*(\*?\*?\w+)\s*(\(.*?\))?\s*:\s*(.+)").ok()?;
```

Comment:
Building the regex-automate is somewhat expensive. That's why regex should be instantiated exactly once and then be reused.

https://docs.rs/regex/latest/regex/#avoid-re-compiling-regexes-especially-in-a-loop

We should also double check if any of the other performance recommendations apply